Classes/EventPublisher/Mapping/DefaultMappingProvider.php

```diff
@@ -0,0 +1,222 @@
+                $presetId = $eventStoreIdentifier . '.' . $pattern;
+                $presets[$presetId] = is_array($presetOptions) ? $presetOptions : [];
+                $presetMatchesAnyListeners = false;
+                foreach ($listeners as $listenerClassName => $events) {
+                    if (preg_match('/^' . str_replace('\\', '\\\\', $pattern) . '$/', $listenerClassName) !== 1) {
+                        continue;
+                    }
+                    if (isset($matchedListeners[$listenerClassName])) {
+                        if ($eventStoreIdentifier === $matchedListeners[$listenerClassName]['eventStoreIdentifier']) {
+                            $message = 'Listener "%s" matches presets "%s" and "%4$s" of Event Store "%5$s". One of the presets need to be adjusted or removed.';
+                        } else {
+                            $message = 'Listener "%s" matches preset "%s" of Event Store "%s" and preset "%s" of Event Store "%s". One of the presets need to be adjusted or removed.';
+                        }
+                        throw new InvalidConfigurationException(sprintf($message, $listenerClassName, $matchedListeners[$listenerClassName]['pattern'], $matchedListeners[$listenerClassName]['eventStoreIdentifier'], $pattern, $eventStoreIdentifier), 1577532711);
+                    }
+                    $presetMatchesAnyListeners = true;
+                    $matchedListeners[$listenerClassName] = compact('eventStoreIdentifier', 'pattern');
+                    foreach ($events as $eventClassName => $handlerMethodName) {
+                        $mappings[$eventStoreIdentifier][] = compact('eventClassName', 'listenerClassName', 'presetId');
+                    }
+                }
+                if (!$presetMatchesAnyListeners) {
+                    throw new InvalidConfigurationException(sprintf('The pattern "%s" for Event Store "%s" does not match any listeners. Please adjust the pattern or remove it', $pattern, $eventStoreIdentifier), 1577533005);
+                }
+            }
+        }
+        $unmatchedListeners = array_diff_key($listeners, $matchedListeners);
+        if ($unmatchedListeners !== []) {
+            if (count($unmatchedListeners) === 1) {
+                $errorMessage = 'The Event Listener "%s" is not registered to any of the configured Event Stores. Please add a new listener pattern at Neos.EventSourcing.EventStore.stores.*.listeners or remove the unused listener class';
+            } else {
+                $errorMessage = 'The following Event Listeners are not registered to any of the configured Event Stores: "%s". Please add new listener patterns at Neos.EventSourcing.EventStore.stores.*.listeners or remove the unused listener classes';
+            }
+            throw new InvalidConfigurationException(sprintf($errorMessage, implode('", "', array_keys($unmatchedListeners))), 1577532358);
+        }
+        return compact('mappings', 'presets');
```

Comment:
uuhm, can we do return [$mappings, $presets] here?

Reply:
Mh, but why?
Isn't `$mappingsAndOptions['presets']` easier to comprehend than `$mappingsAndOptions[1]`?
Would it help to replace this with
```php
return ['mappings' => $mappings, 'presets' => $presets];
```
or why do you prefer a numeric array here?

Reply:
ah return ['mappings' => $mappings, 'presets' => $presets]; is perfect 👯‍♂

Reply:
That's exactly what  `compact('mappings', 'presets');` does